### PR TITLE
add mem

### PR DESCRIPTION
**Pull request by SamVerschueren**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `npm run lint -- package-name` if a `tslint.json` is present.

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Run `tsc` without errors.
- [x] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

### CODE REVIEW

File: mem/mem.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// Type definitions for mem v1.1.0
```
Comment (andy-ms) [inline mem/mem.d.ts]: Delete this file, use `index.d.ts`.

File: mem/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+// Type definitions for mem 1.1.0
+// Project: https://github.com/sindresorhus/mem
+// Definitions by: Sam Verschueren <https://github.com/SamVerschueren>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+declare module "mem" {
```
Comment (andy-ms) [inline mem/index.d.ts]: Don't need this wrapper. (Do it like you did `mem.d.ts`.)

File: mem/mem-tests.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+/// <reference path="mem.d.ts" />
```
Comment (andy-ms) [inline mem/mem-tests.ts]: Don't need this reference.

File: mem/mem-tests.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import * as mem from 'mem';
```
Comment (andy-ms) [inline mem/mem-tests.ts]: Tests won't actually run, so you just need to test the validity of types. e.g.:
```ts
function f(x: number, y: string): number { return x; }
mem(f)(1, "1");
```

File: mem/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+	interface Options {
+		maxAge?: number;
+		cacheKey?: (...args: any[]) => string;
+		cache?: any;
+	}
+
+	function mem<T>(fn: (...args: any[]) => T, options?: Options): (...args: any[]) => T;
```
Comment (andy-ms) [inline mem/index.d.ts]: Could this be `function mem<F extends Function>(fn: F, options?: Options): F;`?

File: mem/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+	namespace mem { }
```
Comment (andy-ms) [inline mem/index.d.ts]: We don't recommend doing this trick; see [here](https://stackoverflow.com/questions/39415661/why-cant-i-import-a-class-or-function-with-import-as-x-from-y).

File: mem/index.d.ts (hunk 5 above)
Comment (SamVerschueren) [inline mem/index.d.ts]: Good point, not sure what would fit the best. What your declaration, you would have to write something like this.

```js
const memGot = mem<(url: string) => Promise<{body: any}>>(got);
```

With my declaration, it looks like this

```js
const memGot = mem<Promise<{body: any}>>(got);
```

But I don't have control over the parameters of the function while with yours, you have. So yes, yours might be slightly better.

**Comment by SamVerschueren** [comment]
@andy-ms Thanks for looking into it. I know about the `namespace` trick although I might not fully understand the problem with it for the future. It's just that a lot of modules are defined this way which makes it hard to use of half of them use the CJS syntax and the other half uses the ESM syntax. What's the policy at DefinitelyTyped on this? Don't allow it or just make sure people understand it could cause issues in the future?

**Comment by sandersn** [comment]
The problem with the namespace trick is that it allows you to write ES module syntax: `import * as foo from 'bar'` but it *only* produces the right Javascript emit if you downlevel to CommonJS modules. If you have TypeScript emit ES6-compliant modules, it just emits `import * as foo from 'bar'`, but then `foo` won't actually be callable. And TypeScript has dedicated syntax for CommonJS: `import foo = require('bar')`. The DefinitelyTyped readme [mentions it](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes) and then links to the stack overflow question that @andy-ms did.

**Comment by SamVerschueren** [comment]
Thanks for the feedback everyone. I removed the namespace hack.

**Comment by SamVerschueren** [comment]
Any idea why this isn't available yet at `@types/mem`? Or does it usually takes longer to publish?

**Comment by andy-ms** [comment]
Having some problems on our server but I published manually, so it's up now.

**Pull request merged**